modules/geochemistry/src/utils/EquilibriumGeochemicalSystem.C

```diff
@@ -0,0 +1,1310 @@
+                                                 unsigned & num_basis_in_algebraic_system,
+                                                 unsigned & num_in_algebraic_system,
+                                                 std::vector<unsigned> & algebraic_index,
+                                                 std::vector<unsigned> & basis_index) const
+{
+  // build in_algebraic_system
+  for (const auto & name_index : _mgd.basis_species_index)
+  {
+    const std::string name = name_index.first;
+    const unsigned basis_ind = name_index.second;
+    const ConstraintMeaningEnum meaning = _constraint_meaning[basis_ind];
+    if (name == "H2O")
+      in_algebraic_system[basis_ind] = (meaning == ConstraintMeaningEnum::MOLES_BULK_WATER);
+    else if (_mgd.basis_species_gas[basis_ind])
+      in_algebraic_system[basis_ind] = false;
+    else if (_mgd.basis_species_mineral[basis_ind])
+      in_algebraic_system[basis_ind] = false;
+    else
+      in_algebraic_system[basis_ind] = (meaning == ConstraintMeaningEnum::MOLES_BULK_SPECIES);
+  }
+
+  // build algebraic_index and basis_index
+  num_basis_in_algebraic_system = 0;
+  algebraic_index.resize(_num_basis, 0);
+  basis_index.resize(_num_basis, 0);
+  for (unsigned basis_ind = 0; basis_ind < _num_basis; ++basis_ind)
+    if (in_algebraic_system[basis_ind])
+    {
+      algebraic_index[basis_ind] = _num_basis_in_algebraic_system;
+      basis_index[_num_basis_in_algebraic_system] = basis_ind;
+      num_basis_in_algebraic_system += 1;
+    }
+
+  num_in_algebraic_system = num_basis_in_algebraic_system + _num_surface_pot;
+}
+
+unsigned
+EquilibriumGeochemicalSystem::getNumInAlgebraicSystem() const
+{
+  return _num_in_algebraic_system;
+}
+
+unsigned
+EquilibriumGeochemicalSystem::getNumBasisInAlgebraicSystem() const
+{
+  return _num_basis_in_algebraic_system;
+}
+
+unsigned
+EquilibriumGeochemicalSystem::getNumSurfacePotentials() const
+{
+  return _num_surface_pot;
+}
+
+const std::vector<bool> &
+EquilibriumGeochemicalSystem::getInAlgebraicSystem() const
+{
+  return _in_algebraic_system;
+}
+
+const std::vector<unsigned> &
+EquilibriumGeochemicalSystem::getBasisIndexOfAlgebraicSystem() const
+{
+  return _basis_index;
+}
+
+const std::vector<unsigned> &
+EquilibriumGeochemicalSystem::getAlgebraicIndexOfBasisSystem() const
+{
+  return _algebraic_index;
+}
+
+std::vector<Real>
+EquilibriumGeochemicalSystem::getAlgebraicVariableValues() const
+{
+  std::vector<Real> var;
+  for (unsigned a = 0; a < _num_basis_in_algebraic_system; ++a)
+    var.push_back(_basis_molality[_basis_index[a]]);
+  for (unsigned s = 0; s < _num_surface_pot; ++s)
+    var.push_back(_surface_pot_expr[s]);
+  return var;
+}
+
+std::vector<Real>
+EquilibriumGeochemicalSystem::getAlgebraicBasisValues() const
+{
+  std::vector<Real> var;
+  for (unsigned a = 0; a < _num_basis_in_algebraic_system; ++a)
+    var.push_back(_basis_molality[_basis_index[a]]);
+  return var;
+}
```

Comment:
This is the kind of change that I meant, by the way:
```suggestion
std::vector<Real>
EquilibriumGeochemicalSystem::getAlgebraicBasisValues() const
{
  std::vector<Real> var(_num_basis_in_algebraic_system);
  for (unsigned a = 0; a < _num_basis_in_algebraic_system; ++a)
    var[a] = _basis_molality[_basis_index[a]
  return var;
}
```

Don't worry about it right now, but I wanted to clear up the discussion!

Reply:
Oh, i see, remove the push_back